modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -2708,6 +2971,61 @@ void GDScriptAnalyzer::reduce_binary_op(GDScriptParser::BinaryOpNode *p_binary_o
 	}
 
 	p_binary_op->set_datatype(result);
+	binary_op_nullity_guard(p_binary_op);
+	pop_narrowed_nullables(narrowed_identifiers);
+}
+
+void GDScriptAnalyzer::reduce_coalesce_op(GDScriptParser::BinaryOpNode *p_coalesce_op) {
+	GDScriptParser::DataType left_type = p_coalesce_op->left_operand->get_datatype();
+	GDScriptParser::DataType right_type = p_coalesce_op->right_operand->get_datatype();
+
+	if (left_type.is_variant() || right_type.is_variant()) {
+		return;
+	}
+
+	GDScriptParser::DataType non_nullable_left_type = left_type;
+	non_nullable_left_type.is_nullable = false;
+	if (!is_type_compatible(non_nullable_left_type, right_type, false, p_coalesce_op)) {
+		push_error(vformat(R"(Invalid operands "%s" and "%s" for "??" operator.)", left_type.to_string(), right_type.to_string()), p_coalesce_op);
```

Comment:
This should be configurable rather than a hard error since `Variant` catches everything that gets through the type system. With it being configurable different projects can utilize whatever style suits their productivity.

I set the datatype to `GDScriptParser::DataType::VARIANT` here and commented out the error for testing and it feels good to use rather than being told off.